File: ruby/app/services/rabobank_csv_row_parser.rb
require 'digest/md5'

class RabobankCsvRowParser
  def initialize
    @to_account_number_index = 0
    @date_index = 4
    @amount_index = 6
    @payee_index = 9
    @memo_index = 19
    @automatic_incasso_id_index = 16
    @positive_amount_character = '+'
  end

  def parse(row)
    to_account_number = row[@to_account_number_index]
    date = DateTime.parse(row[@date_index])
    payee = row[@payee_index]
    memo = parse_memo_text_from(row)
    inflow = parse_inflow_from(row)
    outflow = parse_outflow_from(row)
    code = generate_code(to_account_number, date, payee, memo, inflow, outflow)
    
    Transaction.new(
      to_account_number: to_account_number, 
      code: code,
      date: date, 
      payee: payee, 
      memo: memo, 
      inflow: inflow, 
      outflow: outflow
    )
  end

  private

  def parse_memo_text_from(row)
    incasso_id = row[@automatic_incasso_id_index]
    incasso_text = incasso_id.strip.blank? ? '' : " (Incasso: #{incasso_id})"

    "#{row[@memo_index]}#{incasso_text}"
  end

  def parse_inflow_from(row)
    amount = parse_amount_from row

    @is_positive_amount ? amount : nil
  end

  def parse_outflow_from(row)
    amount = parse_amount_from row

    !@is_positive_amount ? amount : nil
  end

  def parse_amount_from(row)
    amount_string = row[@amount_index]
    amount_string = amount_string.gsub(',', '.')
    @is_positive_amount = amount_string[0] == @positive_amount_character
    parseable_amount_string = amount_string[1..-1]

    parseable_amount_string.to_d
  end

  def generate_code(*transaction_params)
    variables = transaction_params.join("|")
    Digest::MD5.hexdigest(variables)
  end
end

File: django/transaction_management/services/rabobank_csv_row_parser.py
from decimal import Decimal
from transaction_management.models import Transaction
from datetime import datetime
from hashlib import md5


class RabobankCsvRowParser:
    to_account_number_index = 0
    date_index = 4
    amount_index = 6
    payee_index = 9
    memo_index = 19
    automatic_incasso_id_index = 16
    positive_amount_character = '+'

    def parse(self, row: list) -> Transaction:
        amount_str = self._get_amount_str(row)
        is_positive_amount = amount_str[0] == self.positive_amount_character
        amount = Decimal(amount_str[1:])
        transaction = Transaction(
            to_account_number=row[self.to_account_number_index],
            date=datetime.strptime(row[self.date_index], '%Y-%m-%d'),
            payee=row[self.payee_index],
            memo=self._parse_memo_text_from(row),
            inflow=self._parse_inflow_from(amount, is_positive_amount),
            outflow=self._parse_outflow_from(amount, is_positive_amount),
        )

        transaction.code = self._create_unique_code(transaction)
        return transaction

    def _get_amount_str(self, row):
        return row[self.amount_index].replace(',', '.')

    def _parse_memo_text_from(self, row) -> str:
        memo = row[self.memo_index]
        incasso_id = row[self.automatic_incasso_id_index]
        incasso_text = '' if not incasso_id.strip(
        ) else f" (Incasso: {incasso_id})"

        return f"{memo}{incasso_text}"

    def _parse_inflow_from(self, amount, is_positive_amount):
        return amount if is_positive_amount else None

    def _parse_outflow_from(self, amount, is_positive_amount):
        return amount if not is_positive_amount else None

    def _create_unique_code(self, transaction):
        attrs = (
            transaction.date,
            transaction.to_account_number,
            transaction.inflow,
            transaction.outflow,
            transaction.payee,
            transaction.memo)

        attrs_joined = '|'.join([str(attr)
                                 for attr in attrs if attr is not None])
        return md5('|'.join(attrs_joined).encode('utf-8')).hexdigest()
